server/xcode/xcode_darwin.go

```diff
@@ -0,0 +1,171 @@
+	}
+	return xv.developerDirPath, nil
+}
+
+// Return true if the given SDK path is present in the given XCode version.
+func (x *xcodeLocator) IsSDKPathPresentForVersion(sdkPath, version string) bool {
+	xv := x.xcodeVersionForVersionString(version)
+	if xv == nil {
+		return false
+	}
+	for _, sdk := range xv.sdks {
+		if sdk == sdkPath {
+			return true
+		}
+	}
+	return false
+}
+
+// Returns the xcodeVersion most closely matching the version string.
+func (x *xcodeLocator) xcodeVersionForVersionString(version string) *xcodeVersion {
+	versionComponents := strings.Split(version, ".")
+	for i := range versionComponents {
+		subVersion := strings.Join(versionComponents[0:len(versionComponents)-i], ".")
+		if xcodeVersion, ok := x.versions[subVersion]; ok {
+			return xcodeVersion
+		}
+	}
+	return nil
+}
+
+// Locates all all XCode versions installed on the host machine.
+// Very losely based on https://github.com/bazelbuild/bazel/blob/master/tools/osx/xcode_locator.m
+func (x *xcodeLocator) locate() {
+	bundleID := stringToCFString(xcodeBundleID)
+	defer C.CFRelease(C.CFTypeRef(bundleID))
+
+	urlsPointer := C.LSCopyApplicationURLsForBundleIdentifier(bundleID, nil)
+	defer C.CFRelease(C.CFTypeRef(urlsPointer))
+
+	urlsArrayRef := C.CFArrayRef(urlsPointer)
+	n := C.CFArrayGetCount(urlsArrayRef)
+
+	urlRefs := make([]C.CFURLRef, n)
+	C.CFArrayGetValues(urlsArrayRef, C.CFRange{0, n}, (*unsafe.Pointer)(unsafe.Pointer(&urlRefs[0])))
+
+	versionMap := make(map[string]*xcodeVersion)
+	for _, urlRef := range urlRefs {
+		path := "/" + strings.TrimLeft(stringFromCFString(C.CFURLGetString(C.CFURLRef(urlRef))), filePrefix)
+
+		versionFileReader, err := os.Open(path + versionPlistPath)
+		if err != nil {
+			log.Warningf("Error reading version file for XCode located at %s: %s", path, err.Error())
+			continue
+		}
+
+		// The interesting bits to pull from XCode's version plist.
+		var xcodePlist struct {
+			CFBundleShortVersionString string `plist:"CFBundleShortVersionString"`
+			ProductBuildVersion        string `plist:"ProductBuildVersion"`
+		}
+		if err := plist.NewXMLDecoder(versionFileReader).Decode(&xcodePlist); err != nil {
+			log.Warningf("Error decoding plist for XCode located at %s: %s", path, err.Error())
+			continue
+		}
+
+		developerDirPath := path + developerDirectoryPath
+
+		sdks, err := fs.Glob(os.DirFS(developerDirPath), "Platforms/*.platform/Developer/SDKs/*")
+		if err != nil {
+			log.Warningf("Error reading XCode SDKs from %s: %s", path, err.Error())
+			continue
+		}
+
+		versions := expandXCodeVersions(xcodePlist.CFBundleShortVersionString, xcodePlist.ProductBuildVersion)
+		mostPreciseVersion := versions[len(versions)-1]
+		for _, version := range versions {
+			existingXcode, ok := versionMap[version]
+			if ok && mostPreciseVersion < existingXcode.version {
+				continue
+			}
+			log.Infof("Found XCode version: %s=>%s", version, developerDirPath)
+			log.Debugf("With SDKs %+v", sdks)
```

Comment:
can you consolidate this into the line above (or rm it)?

Reply:
Done